[jit] add dynamic isinstance

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26271 [JIT] add type refinements for isinstance checks
* #26270 [jit] refactor extra sugared values
* **#26269 [jit] add dynamic isinstance**
* #26268 Allow types as node attributes
* #26145 [JIT] add CondValue to unify refinements and code emission
* #26061 [jit] Refactor emitIsInstance
* #25447 Use static type information to restore type tags
* #25439 Add ivalue::type(), part 1
* #25418 Fix BUILD TEST

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we do not rely 
on overload support for all instances of polymorphic arguments.

Differential Revision: [D17412853](https://our.internmc.facebook.com/intern/diff/D17412853)

suo commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -264,7 +264,7 @@ RegisterOperators reg(
      Operator(
          "prim::BailOut(...) -> Tensor(a)",
-         [](const Node* /* node */) -> Operation {
+         [](const Node * /* node */) -> Operation {
```
nit: looks like some random formatting changes got through

suo reviewed (approved):
ngnt. I kind of hate the "list", "tuple" thing but I guess that means I kind of hate Python.

zdevito committed:
Update on "[jit] add dynamic isinstance"

[jit] add dynamic isinstance

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we can
remove our __overload__ support.it

gh-metadata: pytorch pytorch 26269 gh/zdevito/109/head

zdevito committed:
Update on "[jit] add dynamic isinstance"

[jit] add dynamic isinstance

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we can
remove our __overload__ support.it

gh-metadata: pytorch pytorch 26269 gh/zdevito/109/head

zdevito committed:
Update on "[jit] add dynamic isinstance"

[jit] add dynamic isinstance

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we can
remove our __overload__ support.it

gh-metadata: pytorch pytorch 26269 gh/zdevito/109/head

zdevito committed:
Update on "[jit] add dynamic isinstance"

[jit] add dynamic isinstance

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we can
remove our __overload__ support.it

gh-metadata: pytorch pytorch 26269 gh/zdevito/109/head

zdevito committed:
Update on "[jit] add dynamic isinstance"

[jit] add dynamic isinstance

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we can
remove our __overload__ support.it

gh-metadata: pytorch pytorch 26269 gh/zdevito/109/head

zdevito committed:
Update on "[jit] add dynamic isinstance"

[jit] add dynamic isinstance

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we can
remove our __overload__ support.it

gh-metadata: pytorch pytorch 26269 gh/zdevito/109/head

zdevito committed:
Update on "[jit] add dynamic isinstance"

[jit] add dynamic isinstance

previously isinstance only worked when we could statically determine
if it were true/false. Now we actually can issue an isinstance check
in case where it is dependent on the runtime type, e.g. Optional[int]
being an instance of int. This is not very useful on its own yet,
but with type refinement and allowing Any as an argument type this will
allow for python-style "overloaded" functions such that we can
remove our __overload__ support.it

gh-metadata: pytorch pytorch 26269 gh/zdevito/109/head
